File: scripts/cache_source_preview_images.py
#!/usr/bin/env python3
"""Cache preview images resolved from recipe source URLs.

This script visits each recipe's ``sourceUrl``, attempts to extract a preview
image (``og:image`` and related meta tags), downloads the image into the local
``public/images/recipes`` directory, and rewrites the recipe's ``imageUrl`` to
reference the cached copy.
"""

from __future__ import annotations

import argparse
import hashlib
import html
import json
import mimetypes
import re
import sys
from pathlib import Path
from typing import Dict, Iterable, Optional, Tuple
from urllib.parse import urljoin, urlparse

import requests


DEFAULT_RECIPES_PATH = Path("recipes.json")
DEFAULT_OUTPUT_PATH = DEFAULT_RECIPES_PATH
DEFAULT_IMAGE_DIR = Path("public/images/recipes")
USER_AGENT = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
    "AppleWebKit/537.36 (KHTML, like Gecko) "
    "Chrome/118.0 Safari/537.36"
)

META_CANDIDATES: Tuple[Tuple[str, str], ...] = (
    ("property", "og:image:secure_url"),
    ("property", "og:image:url"),
    ("property", "og:image"),
    ("name", "og:image"),
    ("property", "twitter:image"),
    ("name", "twitter:image"),
    ("property", "twitter:image:src"),
    ("name", "twitter:image:src"),
    ("itemprop", "image"),
    ("name", "thumbnail"),
)

SAFE_SLUG_RE = re.compile(r"[^a-z0-9]+")

PAGE_HEADER_VARIANTS = (
    {
        "User-Agent": USER_AGENT,
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        "Accept-Language": "en-US,en;q=0.9",
        "Pragma": "no-cache",
        "Cache-Control": "no-cache",
    },
    {
        "User-Agent": USER_AGENT,
    },
    None,
)

IMAGE_HEADERS = {
    "User-Agent": USER_AGENT,
    "Accept": "image/avif,image/webp,image/apng,image/*,*/*;q=0.8",
    "Accept-Language": "en-US,en;q=0.9",
    "Pragma": "no-cache",
    "Cache-Control": "no-cache",
}


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--recipes-file",
        type=Path,
        default=DEFAULT_RECIPES_PATH,
        help="Path to recipes JSON file (default: recipes.json)",
    )
    parser.add_argument(
        "--output-file",
        type=Path,
        default=DEFAULT_OUTPUT_PATH,
        help="Where to write the updated JSON (default: overwrite recipes file)",
    )
    parser.add_argument(
        "--image-dir",
        type=Path,
        default=DEFAULT_IMAGE_DIR,
        help="Directory to write cached images (default: public/images/recipes)",
    )
    parser.add_argument(
        "--skip-existing",
        action="store_true",
        help="Reuse existing cached files when available",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Describe planned actions without downloading or writing files",
    )
    return parser.parse_args()


def load_recipes(path: Path) -> Dict:
    with path.open("r", encoding="utf-8") as handle:
        return json.load(handle)


def ensure_directory(path: Path) -> None:
    path.mkdir(parents=True, exist_ok=True)


def slugify(value: str) -> str:
    slug = SAFE_SLUG_RE.sub("-", value.lower()).strip("-")
    if slug:
        return slug[:60]
    return f"recipe-{hashlib.sha1(value.encode('utf-8')).hexdigest()[:10]}"


def guess_extension(url: str, content_type: Optional[str]) -> str:
    parsed = urlparse(url)
    suffix = Path(parsed.path).suffix.lower()
    if suffix and not suffix.endswith(".php"):
        if suffix in {".jpg", ".jpeg", ".png", ".gif", ".webp", ".avif"}:
            return ".jpg" if suffix == ".jpeg" else suffix

    if content_type:
        ctype = content_type.split(";")[0].strip()
        guess = mimetypes.guess_extension(ctype)
        if guess:
            if guess == ".jpe":
                return ".jpg"
            return guess

    return ".jpg"


def to_public_path(path: Path) -> Optional[str]:
    try:
        relative = path.relative_to(Path("public"))
    except ValueError:
        return None
    return f"/{relative.as_posix()}"


def _meta_regex(attr: str, value: str) -> re.Pattern[str]:
    pattern = rf'<meta[^>]*{attr}\s*=\s*(["\']){re.escape(value)}\1[^>]*>'
    return re.compile(pattern, re.IGNORECASE)


def _extract_content(tag: str) -> Optional[str]:
    match = re.search(r'content\s*=\s*(["\'])(.*?)\1', tag, re.IGNORECASE)
    if not match:
        return None
    return html.unescape(match.group(2).strip())


def _extract_preview_from_html(html_text: str, base_url: str) -> Optional[str]:
    snippet = html_text[:200_000]

    for attr, value in META_CANDIDATES:
        match = _meta_regex(attr, value).search(snippet)
        if match:
            content = _extract_content(match.group(0))
            if content:
                return urljoin(base_url, content)

    link_match = re.search(
        r'<link[^>]*rel\s*=\s*(["\'])(?:image_src|thumbnail)\1[^>]*>',
        snippet,
        re.IGNORECASE,
    )
    if link_match:
        href_match = re.search(r'href\s*=\s*(["\'])(.*?)\1', link_match.group(0), re.IGNORECASE)
        if href_match:
            return urljoin(base_url, html.unescape(href_match.group(2).strip()))

    return None


def resolve_preview_url(source_url: str) -> Tuple[Optional[str], Optional[str]]:
    last_final_url: Optional[str] = None
    for headers in PAGE_HEADER_VARIANTS:
        try:
            response = requests.get(source_url, headers=headers, timeout=30)
            response.raise_for_status()
        except requests.RequestException as exc:
            last_final_url = source_url
            print(f"[warning] Failed to fetch {source_url}: {exc}", file=sys.stderr)
            continue

        final_url = response.url or source_url
        last_final_url = final_url
        preview_url = _extract_preview_from_html(response.text, final_url)
        if preview_url:
            return preview_url, final_url

    return None, last_final_url


def download_image(url: str, referer: Optional[str]) -> Tuple[bytes, Optional[str]]:
    headers = dict(IMAGE_HEADERS)
    if referer:
        headers["Referer"] = referer
    try:
        response = requests.get(url, headers=headers, timeout=30)
        response.raise_for_status()
        return response.content, response.headers.get("Content-Type")
    except requests.RequestException as exc:
        raise RuntimeError(str(exc)) from exc


def update_recipes(
    recipes: Dict,
    image_dir: Path,
    skip_existing: bool,
    dry_run: bool,
) -> bool:
    ensure_directory(image_dir)
    changed = False
    slug_counts: Dict[str, int] = {}

    for entry in recipes.get("recipes", []):
        source_url = (entry.get("sourceUrl") or "").strip()
        title = (entry.get("title") or "recipe").strip() or "recipe"

        if not source_url:
            continue

        slug = slugify(title)
        slug_counts[slug] = slug_counts.get(slug, 0) + 1
        final_slug = slug if slug_counts[slug] == 1 else f"{slug}-{slug_counts[slug]}"

        existing_local = (entry.get("imageUrl") or "").strip().startswith("/")
        if skip_existing and existing_local:
            continue

        preview_url, referer = resolve_preview_url(source_url)
        candidates: Iterable[Tuple[str, Optional[str], str]] = []
        if preview_url:
            candidates = [(preview_url, referer, "preview")]
        else:
            fallback = (entry.get("imageUrl") or "").strip()
            if fallback:
                candidates = [(fallback, source_url, "fallback imageUrl")]

        saved_path: Optional[Path] = None
        last_error: Optional[str] = None

        for candidate_url, candidate_referer, label in candidates:
            target_ext = guess_extension(candidate_url, None)
            target_path = image_dir / f"{final_slug}{target_ext}"

            if skip_existing and target_path.exists():
                saved_path = target_path
                break

            if dry_run:
                print(f"[dry-run] Would download {candidate_url} ({label}) -> {target_path}")
                saved_path = target_path
                break

            try:
                content, content_type = download_image(candidate_url, candidate_referer)
                target_ext = guess_extension(candidate_url, content_type)
                target_path = image_dir / f"{final_slug}{target_ext}"
                target_path.write_bytes(content)
                print(f"Downloaded {candidate_url} ({label}) -> {target_path}")
                saved_path = target_path
                break
            except Exception as exc:  # noqa: BLE001
                last_error = str(exc)
                print(f"[warning] Failed to download {candidate_url} ({label}): {exc}", file=sys.stderr)

        if not saved_path:
            if last_error:
                print(f"[warning] No image cached for {source_url}: {last_error}", file=sys.stderr)
            continue

        if dry_run:
            continue

        mapped = to_public_path(saved_path)
        if mapped:
            if entry.get("imageUrl") != mapped:
                entry["imageUrl"] = mapped
                changed = True
        else:
            entry["imageUrl"] = saved_path.as_posix()
            changed = True

    return changed


def main() -> int:
    args = parse_args()

    recipes = load_recipes(args.recipes_file)
    changed = update_recipes(recipes, args.image_dir, args.skip_existing, args.dry_run)

    if args.dry_run:
        return 0

    if changed:
        with args.output_file.open("w", encoding="utf-8") as handle:
            json.dump(recipes, handle, ensure_ascii=False, indent=2)
            handle.write("\n")
        print(f"Wrote updated recipes to {args.output_file}")
    else:
        print("No changes made.")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/cacheSourcePreviewImages.js
#!/usr/bin/env node
/**
 * Fetch preview images from each recipe's source URL, cache them locally,
 * and rewrite the recipe JSON to point at the cached copies.
 *
 * Usage:
 *   node scripts/cacheSourcePreviewImages.js [options]
 *
 * Options:
 *   --recipes-file=<path>   Path to recipes JSON (default: recipes.json)
 *   --output-file=<path>    Where to write updated JSON (default: overwrite recipes file)
 *   --image-dir=<path>      Directory to store cached images (default: public/images/recipes)
 *   --skip-existing         Reuse existing cached files instead of redownloading
 *   --dry-run               Describe actions without writing files
 */

/* eslint-disable no-console */

import fs from 'node:fs/promises';
import path from 'node:path';
import crypto from 'node:crypto';

const DEFAULT_RECIPES_PATH = path.resolve('recipes.json');
const DEFAULT_OUTPUT_PATH = DEFAULT_RECIPES_PATH;
const DEFAULT_IMAGE_DIR = path.resolve('public/images/recipes');
const USER_AGENT =
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) ' +
  'AppleWebKit/537.36 (KHTML, like Gecko) ' +
  'Chrome/118.0 Safari/537.36';
const VALID_EXTENSIONS = ['.jpg', '.jpeg', '.png', '.gif', '.webp'];
const META_CANDIDATES = [
  { attr: 'property', value: 'og:image:secure_url' },
  { attr: 'property', value: 'og:image:url' },
  { attr: 'property', value: 'og:image' },
  { attr: 'name', value: 'og:image' },
  { attr: 'property', value: 'twitter:image' },
  { attr: 'name', value: 'twitter:image' },
  { attr: 'property', value: 'twitter:image:src' },
  { attr: 'name', value: 'twitter:image:src' },
  { attr: 'itemprop', value: 'image' },
  { attr: 'name', value: 'thumbnail' }
];

function parseArgs(argv) {
  const options = {
    recipesFile: DEFAULT_RECIPES_PATH,
    outputFile: DEFAULT_OUTPUT_PATH,
    imageDir: DEFAULT_IMAGE_DIR,
    skipExisting: false,
    dryRun: false
  };

  argv.slice(2).forEach((arg) => {
    if (arg === '--skip-existing') {
      options.skipExisting = true;
    } else if (arg === '--dry-run') {
      options.dryRun = true;
    } else if (arg.startsWith('--recipes-file=')) {
      options.recipesFile = path.resolve(arg.split('=')[1]);
    } else if (arg.startsWith('--output-file=')) {
      options.outputFile = path.resolve(arg.split('=')[1]);
    } else if (arg.startsWith('--image-dir=')) {
      options.imageDir = path.resolve(arg.split('=')[1]);
    } else if (arg === '--help' || arg === '-h') {
      console.log(
        'Usage: node scripts/cacheSourcePreviewImages.js [options]\n' +
          'Options:\n' +
          '  --recipes-file=<path>   Path to recipes JSON (default: recipes.json)\n' +
          '  --output-file=<path>    Where to write updated JSON (default: overwrite recipes file)\n' +
          '  --image-dir=<path>      Directory for cached images (default: public/images/recipes)\n' +
          '  --skip-existing         Reuse existing cached files instead of redownloading\n' +
          '  --dry-run               Describe actions without writing files\n'
      );
      process.exit(0);
    }
  });

  return options;
}

async function ensureDir(dirPath) {
  await fs.mkdir(dirPath, { recursive: true });
}

async function loadJson(filePath) {
  const raw = await fs.readFile(filePath, 'utf8');
  return JSON.parse(raw);
}

function slugify(value) {
  const lower = value.toLowerCase();
  const slug = lower.replace(/[^a-z0-9]+/g, '-').replace(/(^-|-$)/g, '');
  if (slug) {
    return slug.slice(0, 60);
  }
  return `recipe-${crypto.createHash('sha1').update(value).digest('hex').slice(0, 10)}`;
}

function guessExtension(url, contentType) {
  try {
    const parsed = new URL(url);
    const fromPath = path.extname(parsed.pathname).toLowerCase();
    if (VALID_EXTENSIONS.includes(fromPath)) {
      return fromPath;
    }
  } catch {
    // ignore malformed URLs
  }

  if (contentType) {
    const baseType = contentType.split(';')[0].trim();
    const lookup = {
      'image/jpeg': '.jpg',
      'image/jpg': '.jpg',
      'image/png': '.png',
      'image/gif': '.gif',
      'image/webp': '.webp'
    };
    if (lookup[baseType]) {
      return lookup[baseType];
    }
  }

  return '.jpg';
}

function toPublicPath(filePath) {
  const publicRoot = path.resolve('public');
  const absolute = path.resolve(filePath);
  if (!absolute.startsWith(publicRoot)) {
    return null;
  }
  return `/${path.relative(publicRoot, absolute).replace(/\\/g, '/')}`;
}

async function fileExists(filePath) {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}

async function findExistingForSlug(imageDir, slug) {
  for (const ext of VALID_EXTENSIONS) {
    const candidate = path.join(imageDir, `${slug}${ext}`);
    if (await fileExists(candidate)) {
      return candidate;
    }
  }
  return null;
}

async function fetchText(url, extraHeaders = {}) {
  const response = await fetch(url, {
    redirect: 'follow',
    headers: {
      'User-Agent': USER_AGENT,
      Accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
      'Accept-Language': 'en-US,en;q=0.9',
      Pragma: 'no-cache',
      'Cache-Control': 'no-cache',
      ...extraHeaders
    }
  });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }

  const text = await response.text();
  return { text, finalUrl: response.url };
}

async function fetchBinary(url, referer) {
  const response = await fetch(url, {
    redirect: 'follow',
    headers: {
      'User-Agent': USER_AGENT,
      Accept: 'image/avif,image/webp,image/apng,image/*,*/*;q=0.8',
      'Accept-Language': 'en-US,en;q=0.9',
      Pragma: 'no-cache',
      'Cache-Control': 'no-cache',
      ...(referer ? { Referer: referer } : {})
    }
  });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }

  const arrayBuffer = await response.arrayBuffer();
  return {
    buffer: Buffer.from(arrayBuffer),
    contentType: response.headers.get('content-type') || undefined
  };
}

function decodeHtmlEntities(value) {
  return value
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'");
}

function extractMetaContent(tag) {
  const contentMatch = tag.match(/content\s*=\s*(['"])(.*?)\1/i);
  if (!contentMatch) {
    return null;
  }
  return decodeHtmlEntities(contentMatch[2].trim());
}

function extractPreviewUrl(html, baseUrl) {
  const searchSpace = html.slice(0, 200000);

  for (const { attr, value } of META_CANDIDATES) {
    const regex = new RegExp(`<meta[^>]*${attr}\\s*=\\s*(['"])${value}\\1[^>]*>`, 'i');
    const match = regex.exec(searchSpace);
    if (match) {
      const url = extractMetaContent(match[0]);
      if (url) {
        return resolveUrl(url, baseUrl);
      }
    }
  }

  const linkMatch = searchSpace.match(
    /<link[^>]*rel\s*=\s*(['"])(?:image_src|thumbnail)\1[^>]*>/i
  );
  if (linkMatch) {
    const hrefMatch = linkMatch[0].match(/href\s*=\s*(['"])(.*?)\1/i);
    if (hrefMatch) {
      return resolveUrl(decodeHtmlEntities(hrefMatch[2].trim()), baseUrl);
    }
  }

  return null;
}

function resolveUrl(candidate, base) {
  try {
    return new URL(candidate, base).toString();
  } catch {
    return null;
  }
}

function ensureTrailingSlash(value) {
  return value.endsWith('/') ? value : `${value}/`;
}

async function resolveInstagramPreview(sourceUrl) {
  const normalized = ensureTrailingSlash(sourceUrl.split('?')[0]);
  const oembedUrl = `https://www.instagram.com/oembed/?omitscript=true&url=${encodeURIComponent(normalized)}`;

  try {
    const response = await fetch(oembedUrl, {
      headers: {
        'User-Agent': USER_AGENT,
        Accept: 'application/json',
        'Accept-Language': 'en-US,en;q=0.9'
      }
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }

    const payload = await response.json();
    const thumbnail = (payload.thumbnail_url || '').trim();
    if (thumbnail) {
      return {
        url: thumbnail,
        referer: 'https://www.instagram.com/'
      };
    }
  } catch (error) {
    console.warn(`[warning] Instagram oEmbed failed for ${sourceUrl}: ${error.message}`);
  }

  return null;
}

async function resolvePreviewFromSource(sourceUrl) {
  let parsed;
  try {
    parsed = new URL(sourceUrl);
  } catch {
    parsed = null;
  }

  if (parsed && parsed.hostname.includes('instagram.com')) {
    const instagramPreview = await resolveInstagramPreview(sourceUrl);
    if (instagramPreview) {
      return instagramPreview;
    }
  }

  try {
    const { text, finalUrl } = await fetchText(sourceUrl);
    const previewUrl = extractPreviewUrl(text, finalUrl || sourceUrl);
    if (!previewUrl) {
      return null;
    }
    return {
      url: previewUrl,
      referer: finalUrl || sourceUrl
    };
  } catch (error) {
    console.warn(`[warning] Failed to fetch preview for ${sourceUrl}: ${error.message}`);
    return null;
  }
}

async function cachePreviews(recipes, options) {
  await ensureDir(options.imageDir);

  let changed = false;
  const slugCounts = new Map();

  for (const recipe of recipes.recipes || []) {
    const sourceUrl = (recipe.sourceUrl || '').trim();
    if (!sourceUrl) {
      continue;
    }

    const title = (recipe.title || 'recipe').trim();
    const baseSlug = slugify(title);
    const count = (slugCounts.get(baseSlug) || 0) + 1;
    slugCounts.set(baseSlug, count);
    const slug = count === 1 ? baseSlug : `${baseSlug}-${count}`;

    if (options.skipExisting) {
      const existing = await findExistingForSlug(options.imageDir, slug);
      if (existing) {
        const mapped = toPublicPath(existing);
        if (mapped && recipe.imageUrl !== mapped) {
          recipe.imageUrl = mapped;
          changed = true;
        }
        continue;
      }
    }

    const candidates = [];
    const preview = await resolvePreviewFromSource(sourceUrl);
    if (preview) {
      candidates.push({
        url: preview.url,
        label: 'source preview',
        referer: preview.referer
      });
    }
    const existingUrl = (recipe.imageUrl || '').trim();
    if (existingUrl && !existingUrl.startsWith('/')) {
      candidates.push({ url: existingUrl, label: 'existing imageUrl', referer: sourceUrl });
    }

    if (candidates.length === 0) {
      continue;
    }

    let savedPath = null;
    for (const candidate of candidates) {
      try {
        if (options.dryRun) {
          console.log(
            `[dry-run] Would download ${candidate.url} (${candidate.label}) -> ${path.join(
              options.imageDir,
              `${slug}<ext>`
            )}`
          );
          savedPath = 'dry-run';
          break;
        }

        const { buffer, contentType } = await fetchBinary(candidate.url, candidate.referer);
        const ext = guessExtension(candidate.url, contentType);
        const targetPath = path.join(options.imageDir, `${slug}${ext}`);
        await fs.writeFile(targetPath, buffer);
        savedPath = targetPath;
        console.log(`Downloaded ${candidate.url} (${candidate.label}) -> ${targetPath}`);
        break;
      } catch (error) {
        console.warn(
          `[warning] Failed to download ${candidate.url} (${candidate.label}): ${error.message}`
        );
      }
    }

    if (!savedPath) {
      continue;
    }

    if (!options.dryRun) {
      const mapped = toPublicPath(savedPath);
      if (mapped) {
        recipe.imageUrl = mapped;
      } else {
        recipe.imageUrl = savedPath;
      }
    }
    changed = true;
  }

  return changed;
}

async function main() {
  const options = parseArgs(process.argv);

  if (typeof fetch !== 'function') {
    console.error('[error] This script requires Node.js 18 or newer (global fetch API).');
    return 1;
  }

  try {
    const recipes = await loadJson(options.recipesFile);
    const changed = await cachePreviews(recipes, options);

    if (options.dryRun) {
      return 0;
    }

    if (changed) {
      const serialized = `${JSON.stringify(recipes, null, 2)}\n`;
      await fs.writeFile(options.outputFile, serialized, 'utf8');
      console.log(`Wrote updated recipes to ${options.outputFile}`);
    } else {
      console.log('No changes made.');
    }
  } catch (error) {
    console.error(`[error] ${error.message}`);
    return 1;
  }

  return 0;
}

main().then(
  (code) => {
    process.exit(code);
  },
  (error) => {
    console.error('[error]', error);
    process.exit(1);
  }
);
